pkg/graph/dag/dag.go

```diff
@@ -136,6 +136,84 @@ func (d *DirectedAcyclicGraph[T]) AddDependencies(from T, dependencies []T) erro
 	return nil
 }
 
+// TopologicalSortLevels returns the vertices of the graph grouped by topological levels
+// using Kahn's algorithm. Each level contains vertices that have no dependencies on each
+// other and can be processed in parallel. Within each level, vertices are sorted by their
+// original Order for stability.
+//
+// For example, given a graph:
+//
+//	A -> C
+//	B -> C
+//	C -> D
+//
+// The result would be: [[A, B], [C], [D]]
+// where A and B can be processed in parallel, then C, then D.
+func (d *DirectedAcyclicGraph[T]) TopologicalSortLevels() ([][]T, error) {
+	// Calculate in-degrees (number of dependencies) for each vertex
+	inDegree := make(map[T]int)
+	for id := range d.Vertices {
+		inDegree[id] = len(d.Vertices[id].DependsOn)
+	}
+
+	// Find all vertices with in-degree 0 (no dependencies)
+	currentLevel := make([]*Vertex[T], 0)
+	for id, degree := range inDegree {
+		if degree == 0 {
+			currentLevel = append(currentLevel, d.Vertices[id])
+		}
+	}
+
+	var levels [][]T
+	processed := 0
+
+	// Process vertices level by level
+	for len(currentLevel) > 0 {
+		// Sort current level by Order to maintain stability and preserve user-provided ordering
+		sort.Slice(currentLevel, func(i, j int) bool {
+			return currentLevel[i].Order < currentLevel[j].Order
+		})
+
+		// Convert vertices to IDs for output
+		levelIDs := make([]T, len(currentLevel))
+		for i, v := range currentLevel {
+			levelIDs[i] = v.ID
+		}
+		levels = append(levels, levelIDs)
+		processed += len(currentLevel)
+
+		// Find next level by decrementing in-degrees
+		nextLevel := make([]*Vertex[T], 0)
+		for _, vertex := range currentLevel {
+			// For each vertex that depends on the current vertex, decrement its in-degree
+			for id, v := range d.Vertices {
+				if _, dependsOn := v.DependsOn[vertex.ID]; dependsOn {
+					inDegree[id]--
+					if inDegree[id] == 0 {
+						nextLevel = append(nextLevel, v)
+					}
+				}
+			}
+		}
+
+		currentLevel = nextLevel
+	}
+
+	// Check if all vertices were processed (if not, there's a cycle)
+	if processed != len(d.Vertices) {
+		hasCycle, cycle := d.hasCycle()
```

Comment:
nit: This is non optimal. Kahn has implicit cycle detection and this is a very expensive op